feat: implement chunk deletion for vector stores

derekhiggins opened the pull request:
Add support for deleting individual chunks from vector stores

- Add abstract remove_chunk() method to EmbeddingIndex base class
- Implement chunk deletion for Faiss provider, SQLite Vec, Milvus, PGVector 
- Placeholder implementations with NotImplementedError for Chroma/Qdrant/Weaviate
- Integrate chunk deletion into OpenAI vector store file deletion flow
- removed xfail from test_openai_vector_store_delete_file_removes_from_vector_store

Closes: #2477

franciscojavierarceo commented on llama_stack/providers/inline/vector_io/sqlite_vec/sqlite_vec.py:
```diff
@@ -426,6 +426,35 @@ async def query_hybrid(
 
         return QueryChunksResponse(chunks=chunks, scores=scores)
 
+    async def remove_chunk(self, chunk_id: str) -> None:
```
```suggestion
    async def delete_chunk(self, chunk_id: str) -> None:
```

derekhiggins replied:
comment below

franciscojavierarceo commented on llama_stack/providers/remote/vector_io/pgvector/pgvector.py:
```diff
@@ -99,7 +99,7 @@ async def add_chunks(self, chunks: list[Chunk], embeddings: NDArray):
             values.append(
                 (
-                    f"{chunk.metadata['document_id']}:chunk-{i}",
+                    f"{chunk.chunk_id}",
```
why use `chunk_id` here and `stored_chunk_id` elsewhere?

franciscojavierarceo reviewed (changes requested):
actually @derekhiggins can you take a look at my most recent changes? I moved `_delete_openai_chunk_from_vector_store` into `openai_vector_store_mixin.py` but definitely deleting the chunks is still important. think you have to make some changes though.

derekhiggins commented:
> actually @derekhiggins can you take a look at my most recent changes? I moved `_delete_openai_chunk_from_vector_store` into `openai_vector_store_mixin.py` but definitely deleting the chunks is still important. think you have to make some changes though.

I'm guessing you mean `_delete_openai_vector_store_file_from_storage` (`_delete_openai_chunk_from_vector_store` is the method I've introduced).

I can't move _delete_openai_chunk_from_vector_store into OpenAIVectorStoreMixin as each one needs access to the internal details of the provider implimentation, so I've instead
renamed it "delete_chunk" and made it an abstract method to be implimented by each provider

(note: I'm on PTO after today for 2 weeks, so if somebody wants to take this over fire ahead, if not I'll pick it up again when I return)

franciscojavierarceo commented:
> > actually @derekhiggins can you take a look at my most recent changes? I moved `_delete_openai_chunk_from_vector_store` into `openai_vector_store_mixin.py` but definitely deleting the chunks is still important. think you have to make some changes though.
> 
> I'm guessing you mean `_delete_openai_vector_store_file_from_storage` (`_delete_openai_chunk_from_vector_store` is the method I've introduced).
> 
> I can't move _delete_openai_chunk_from_vector_store into OpenAIVectorStoreMixin as each one needs access to the internal details of the provider implimentation, so I've instead renamed it "delete_chunk" and made it an abstract method to be implimented by each provider
> 
> (note: I'm on PTO after today for 2 weeks, so if somebody wants to take this over fire ahead, if not I'll pick it up again when I return)

Thanks for clarifying! That makes sense. 

My only nit would be to rename it to `delete_chunks` and instead handle a list of elements to follow the rest of the convention of the provider (`insert_chunks`, `query_chunks`). Let me know if you're still in for the day and if not I can help finish it up.

derekhiggins commented:
> > > actually @derekhiggins can you take a look at my most recent changes? I moved `_delete_openai_chunk_from_vector_store` into `openai_vector_store_mixin.py` but definitely deleting the chunks is still important. think you have to make some changes though.
> > 
> > 
> > I'm guessing you mean `_delete_openai_vector_store_file_from_storage` (`_delete_openai_chunk_from_vector_store` is the method I've introduced).
> > I can't move _delete_openai_chunk_from_vector_store into OpenAIVectorStoreMixin as each one needs access to the internal details of the provider implimentation, so I've instead renamed it "delete_chunk" and made it an abstract method to be implimented by each provider
> > (note: I'm on PTO after today for 2 weeks, so if somebody wants to take this over fire ahead, if not I'll pick it up again when I return)
> 
> Thanks for clarifying! That makes sense.
> 
> My only nit would be to rename it to `delete_chunks` and instead handle a list of elements to follow the rest of the convention of the provider (`insert_chunks`, `query_chunks`). Let me know if you're still in for the day and if not I can help finish it up.

Done, its now `delete_chunks` and takes a list of chunk id's
